### PR TITLE
Smtp frames/v18

### PR DESCRIPTION
**Pull request by victorjulien**
Replace https://github.com/OISF/suricata/pull/11173, rebased

https://redmine.openinfosecfoundation.org/issues/6718
https://redmine.openinfosecfoundation.org/issues/4905

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1868

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 20874

File: src/app-layer-frames.c [hunk 1]
```diff
@@ -83,6 +83,56 @@ static void FrameDebug(const char *prefix, const Frames *frames, const Frame *fr
 #endif
 }
 
+/** \todo perhaps we can search backwards */
```
Comment (catenacyber) [inline src/app-layer-frames.c:86]: This todo looks a good idea

File: src/app-layer-frames.c [hunk 2]
```diff
@@ -83,6 +83,56 @@ static void FrameDebug(const char *prefix, const Frames *frames, const Frame *fr
...
+                candidate = frame;
+        }
+    }
+    return candidate;
+}
+
+/**
+ * \note "open" means a frame that has no length set (len == -1)
+ * \todo perhaps we can search backwards */
+Frame *FrameGetLastOpenByType(Frames *frames, const uint8_t frame_type)
```
Comment (catenacyber) [inline src/app-layer-frames.c:113]: Is this for both directions ?

File: src/app-layer-smtp.c [hunk 3]
```diff
@@ -681,6 +730,11 @@ static AppLayerResult SMTPGetLine(
         input->len -= line->len;
         DEBUG_VALIDATE_BUG_ON((input->consumed + input->len) != input->orig_len);
         line->buf = input->buf + o_consumed;
+
+        if (frame != NULL) {
+            frame->len = (int64_t)line->len;
```
Comment (catenacyber) [inline src/app-layer-smtp.c:735]: Does this work for multiline responses ? Like server initial banner with capabilities...

File: src/app-layer-smtp.c [hunk 4]
```diff
@@ -1334,13 +1388,28 @@ static inline void ResetLine(SMTPLine *line)
...
     DEBUG_VALIDATE_BUG_ON(line->len != 0);
     DEBUG_VALIDATE_BUG_ON(line->delim_len != 0);
 
+    Frame *data_frame = NULL;
+    if (state->curr_tx) {
+        data_frame = state->curr_tx->data_frame_opened
+                             ? AppLayerFrameGetLastOpenByType(f, 0, SMTP_FRAME_DATA)
+                             : NULL;
+        if (data_frame == NULL && !state->curr_tx->data_frame_opened) {
+            data_frame = AppLayerFrameNewByPointer(
```
Comment (catenacyber) [inline src/app-layer-smtp.c:1404]: Is it not too soon to open this frame ?

File: src/app-layer-smtp.h [hunk 5]
```diff
@@ -81,6 +82,8 @@ typedef struct SMTPTransaction_ {
     // another DATA command within the same context
     // will trigger an app-layer event.
     bool is_data;
+    /** did we try to open the data frame? */
+    bool data_frame_opened;
```
Comment (catenacyber) [inline src/app-layer-smtp.h:86]: Could not `is_data` be used ?

File: src/flow.h [hunk 6]
```diff
@@ -117,6 +117,9 @@ typedef struct AppLayerParserState_ AppLayerParserState;
 #define FLOW_TS_APP_UPDATED BIT_U32(29)
 #define FLOW_TC_APP_UPDATED BIT_U32(30)
 
+#define FLOW_TS_APP_UPDATE_NEXT BIT_U32(31)
+#define FLOW_TC_APP_UPDATE_NEXT BIT_U32(2)
```
Comment (catenacyber) [inline src/flow.h:121]: maybe remove the `vacancy` comment and say it is used now ;-)

File: src/flow.h [hunk 7]
```diff
@@ -117,6 +117,9 @@ typedef struct AppLayerParserState_ AppLayerParserState;
 #define FLOW_TS_APP_UPDATED BIT_U32(29)
 #define FLOW_TC_APP_UPDATED BIT_U32(30)
 
+#define FLOW_TS_APP_UPDATE_NEXT BIT_U32(31)
```
Comment (catenacyber) [inline src/flow.h:120]: Could these get a comment ? Not sure I understand the purpose

File: src/stream-tcp-reassemble.c [hunk 8]
```diff
@@ -1983,6 +1983,9 @@ int StreamTcpReassembleHandleSegment(ThreadVars *tv, TcpReassemblyThreadCtx *ra_
         }
     } else if (ssn->state == TCP_CLOSED) {
         dir = UPDATE_DIR_BOTH;
+    } else if ((ssn->flags & STREAMTCP_FLAG_ASYNC) != 0) {
+        dir = UPDATE_DIR_PACKET;
```
Comment (catenacyber) [inline src/stream-tcp-reassemble.c:1987]: Is there a SV test for this change ?

File: src/app-layer-smtp.c (hunk 3 above)
Comment (catenacyber) [reply src/app-layer-smtp.c:735]: Should not
`alert smtp any any -> any any (frame:smtp.response_line; content:"250-PIPELINING"; content:"250-SIZE"; sid:3;)` trigger on smtp-pipelining SV test ?

**Review by catenacyber — COMMENTED** [review]
Thanks for the work :-)

- CI : 🟢 
- Code : Looking good, will reread the SMTP buffering part
- Commits segmentation : ok, maybe first 2 commits ca be squashed together
- Commit messages : ok for me

> detect/frames: avoid IPS rescanning

Is this not to avoid scanning frames when we just get a TCP ACK without any payload ?

- Git ID set : looks fine for me
- CLA : you already contributed
- Doc update : automatic, right ?
- Redmine ticket : ok
- Rustfmt : no rust
- Tests : looking into the SV test
- Dependencies added: none

**Comment by catenacyber** [comment]
You do not seem to set the frame tx_id so we do not get the app-layer metadata logged, do we ?

File: src/app-layer-smtp.c [hunk 9]
```diff
@@ -1396,8 +1478,14 @@ static int SMTPPreProcessCommands(
             line->delim_len = 0;
 
             /* bail if `SMTPProcessRequest` ended the data mode */
-            if ((state->parser_state & SMTP_PARSER_STATE_COMMAND_DATA_MODE) == 0)
+            if ((state->parser_state & SMTP_PARSER_STATE_COMMAND_DATA_MODE) == 0) {
+                if (data_frame == NULL)
+                    data_frame = AppLayerFrameGetLastOpenByType(f, 0, SMTP_FRAME_DATA);
+                if (data_frame) {
+                    data_frame->len = (slice->offset + input->consumed) - data_frame->offset;
```
Comment (catenacyber) [inline src/app-layer-smtp.c:1485]: Are we sure we have some contiguous buffer from the original created frame ?

File: src/flow.h (hunk 6 above)
Comment (victorjulien) [reply src/flow.h:121]: yeah

File: src/flow.h (hunk 7 above)
Comment (victorjulien) [reply src/flow.h:120]: yes

File: src/stream-tcp-reassemble.c (hunk 8 above)
Comment (victorjulien) [reply src/stream-tcp-reassemble.c:1987]: it makes a SV test pass that was broken by other changes in the PR (direction update stuff)

File: src/app-layer-smtp.c (hunk 3 above)
Comment (victorjulien) [reply src/app-layer-smtp.c:735]: It does not. It's a per line frame. I guess a more generic `smtp.response` (like wireshark) could be added that would bundle up lines that belong in single response. Not planning to do that now though.

File: src/app-layer-smtp.c (hunk 4 above)
Comment (victorjulien) [reply src/app-layer-smtp.c:1404]: I don't think so, why do you think it may be?

File: src/app-layer-smtp.c (hunk 9 above)
Comment (victorjulien) [reply src/app-layer-smtp.c:1485]: yes all data from the open is part of data

**Comment by victorjulien** [comment]
replaced by #11222

File: src/app-layer-smtp.c (hunk 3 above)
Comment (catenacyber) [reply src/app-layer-smtp.c:735]: Fair enough, thanks

File: src/app-layer-smtp.c (hunk 4 above)
Comment (catenacyber) [reply src/app-layer-smtp.c:1404]: cf usage of `is_data` ?

File: src/app-layer-smtp.c (hunk 9 above)
Comment (catenacyber) [reply src/app-layer-smtp.c:1485]: But what happens if I stream indefinitely some DATA to a smtp server ? Can I get an infinite frame ?
Would Suricata keep a 1Gbyte buffer somewhere ?

Comment (victorjulien) [reply src/app-layer-smtp.c:1485]: no, it's a streaming buffer

Comment (victorjulien) [reply src/app-layer-smtp.c:1485]: to clarify: it's a pointer into the stream data. So the data kept is managed by the stream engine. Inspection happens in streaming manner, just like with "raw" content stream inspection.

Comment (catenacyber) [reply src/app-layer-smtp.c:1485]: Ok, looks magical to me

**Pull request closed without merging**